extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/context/TransactionContext.java

```diff
@@ -0,0 +1,213 @@
+import javax.enterprise.context.spi.CreationalContext;
+import javax.transaction.Status;
+import javax.transaction.SystemException;
+import javax.transaction.Transaction;
+import javax.transaction.TransactionManager;
+import javax.transaction.TransactionScoped;
+import javax.transaction.TransactionSynchronizationRegistry;
+
+import com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionSynchronizationRegistryImple;
+
+import io.quarkus.arc.*;
+
+/**
+ * {@link javax.enterprise.context.spi.Context} class which defines the {@link TransactionScoped} context.
+ */
+public class TransactionContext implements InjectableContext {
+    // marker object to be put as a key for SynchronizationRegistry to gather all beans created in the scope
+    private static final Object TRANSACTION_CONTEXT_MARKER = new Object();
+
+    private TransactionSynchronizationRegistry transactionSynchronizationRegistry = new TransactionSynchronizationRegistryImple();
+    private TransactionManager transactionManager = com.arjuna.ats.jta.TransactionManager.transactionManager();
+
+    @Override
+    public void destroy() {
+        if (!isActive()) {
+            return;
+        }
+        TransactionContextState contextState = (TransactionContextState) transactionSynchronizationRegistry
+                .getResource(TRANSACTION_CONTEXT_MARKER);
+        if (contextState == null) {
+            return;
+        }
+        contextState.destroy();
+    }
+
+    @Override
+    public void destroy(Contextual<?> contextual) {
+        if (!isActive()) {
+            return;
+        }
+        TransactionContextState contextState = (TransactionContextState) transactionSynchronizationRegistry
+                .getResource(TRANSACTION_CONTEXT_MARKER);
+        if (contextState == null) {
+            return;
+        }
+        contextState.remove(contextual);
+    }
+
+    @Override
+    public ContextState getState() {
+        if (!isActive()) {
+            throw new ContextNotActiveException("No active transaction on the current thread");
+        }
+
+        TransactionContextState contextState = (TransactionContextState) transactionSynchronizationRegistry
+                .getResource(TRANSACTION_CONTEXT_MARKER);
+        if (contextState == null) {
+            return new TransactionContextState<>();
+        }
+        return contextState;
+    }
+
+    @Override
+    public Class<? extends Annotation> getScope() {
+        return TransactionScoped.class;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T> T get(Contextual<T> contextual, CreationalContext<T> creationalContext) {
+        if (!isActive()) {
+            throw new ContextNotActiveException();
+        }
+        if (contextual == null) {
+            return null;
+        }
+
+        ContextInstanceHandle<T> instanceHandle = null;
+        TransactionContextState<T> contextState = (TransactionContextState<T>) transactionSynchronizationRegistry
```

Comment:
Originally you told me that the beans were stored in the registry (not sure how exactly for Weld?).
Since Narayana already count with the possibility of propagating across threads, is it necessary to introduce a `ConcurrentMap` storage here? Doesn't the registry offer some other way to achieve that?
How exactly are the beans stored with Weld?

Reply:
in weld they are stored directly as beans. The registry is created as `Collections.synchronizedMap(...)`.
But here I chose the transactional context as a container which is managed by `CDI/Arc`. My point is that other libraries may use the `TransactionSynchronizationRegistry.put/getResource` (https://docs.oracle.com/javaee/7/api/javax/transaction/TransactionSynchronizationRegistry.html#putResource-java.lang.Object-java.lang.Object-`). I wanted here to touch only the resources which the Arc put into the container. That's why I brought this synchronized container here. I omitted the need to create transaction `Synchronization` callback for each bean.